timesync: Introduce  network time security protocol

ssahani opened the pull request:
```
implements https://tools.ietf.org/html/rfc8915
```

RFE https://github.com/systemd/systemd/issues/9481

ssahani commented:
@yuwata Can you figure how we going to fix the encryption part ?

yuwata commented:
@ssahani I will take a look later.

bluca commented on meson.build:
```diff
@@ -402,6 +402,7 @@ possible_link_flags = [
         '-Wl,-z,relro',
         '-Wl,-z,now',
         '-fstack-protector',
+        '-lgnutls'
```
shouldn't this get automatically added via the dependency below?

ssahani replied:
@yuwata ?

bluca replied:
It will - that's what dependencies does. You can remove this line.

ssahani replied:
Bi this does not fix

bluca replied:
it wasn't about fixing the build, it's an independent issues - library linking flags are added by Meson using pkg-config, they shouldn't be added manually

ssahani commented:
```
06:44:16 systemd-timesyncd@exe/src_timesync_timesyncd-nts-client.c.o: In function `nts_set_auth':
06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:59: undefined reference to `gnutls_prf_rfc5705'
06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:68: undefined reference to `gnutls_prf_rfc5705'
06:44:16 systemd-timesyncd@exe/src_timesync_timesyncd-nts-client.c.o: In function `ntp_extension_parse_aeef':
06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:411: undefined reference to `gnutls_aead_cipher_init'
06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:432: undefined reference to `gnutls_aead_cipher_decrypt'
06:44:16 systemd-timesyncd@exe/src_timesync_timesyncd-nts-client.c.o: In function `nts_tls_connect':
06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:189: undefined reference to `gnutls_priority_init2'
06:44:16 systemd-timesyncd@exe/src_timesync_timesyncd-nts-client.c.o: In function `ntp_extension_encrypt_auth_field':
06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:353: undefined reference to `gnutls_aead_cipher_init'
06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:379: undefined reference to `gnutls_aead_cipher_encrypt'
06:44:16 systemd-timesyncd@exe/src_timesync_timesyncd-manager.c.o: In function `manager_disconnect':
06:44:16 /root/systemd/build/../src/timesync/timesyncd-manager.c:1100: undefined reference to `gnutls_aead_cipher_deinit'
06:44:16 /root/systemd/build/../src/timesync/timesyncd-manager.c:1105: undefined reference to `gnutls_aead_cipher_deinit'
06:44:16 collect2: error: ld returned 1 exit status
```

The build is failing in centos . Not sure how to fix this.

ssahani commented:
I am moving it for initial review .

wingel reviewed (commented):
Question after a quick glance.  Doesn't this assume that the field is \0-terminated?  If so, that assumption is not correct.

                       d = memdup(data, data_size);
                        if (!d)
                                return -ENOMEM;

                        r = dns_name_is_valid_or_address(d);

If you want to test against a server that uses the NTPV4 Server Negotiation record you can try Netnod's NTS server at nts.netnod.se:4460.  Since I'm the guy responsible for setting up the server at Netnod, please mail me at <wingel@netnod.se> if you have any issues with it and I'll try to help out as best as I can.

bluca commented:
> ```
> 06:44:16 systemd-timesyncd@exe/src_timesync_timesyncd-nts-client.c.o: In function `nts_set_auth':
> 06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:59: undefined reference to `gnutls_prf_rfc5705'
> 06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:68: undefined reference to `gnutls_prf_rfc5705'
> 06:44:16 systemd-timesyncd@exe/src_timesync_timesyncd-nts-client.c.o: In function `ntp_extension_parse_aeef':
> 06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:411: undefined reference to `gnutls_aead_cipher_init'
> 06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:432: undefined reference to `gnutls_aead_cipher_decrypt'
> 06:44:16 systemd-timesyncd@exe/src_timesync_timesyncd-nts-client.c.o: In function `nts_tls_connect':
> 06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:189: undefined reference to `gnutls_priority_init2'
> 06:44:16 systemd-timesyncd@exe/src_timesync_timesyncd-nts-client.c.o: In function `ntp_extension_encrypt_auth_field':
> 06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:353: undefined reference to `gnutls_aead_cipher_init'
> 06:44:16 /root/systemd/build/../src/timesync/timesyncd-nts-client.c:379: undefined reference to `gnutls_aead_cipher_encrypt'
> 06:44:16 systemd-timesyncd@exe/src_timesync_timesyncd-manager.c.o: In function `manager_disconnect':
> 06:44:16 /root/systemd/build/../src/timesync/timesyncd-manager.c:1100: undefined reference to `gnutls_aead_cipher_deinit'
> 06:44:16 /root/systemd/build/../src/timesync/timesyncd-manager.c:1105: undefined reference to `gnutls_aead_cipher_deinit'
> 06:44:16 collect2: error: ld returned 1 exit status
> ```
> 
> The build is failing in centos . Not sure how to fix this.

Those functions have been added in 3.3.4, but Centos7 has an older version of gnutls. You have 2 options: either check for each of those functions manually in Meson, or simply check which version is found and add an additional conf.set10 if it's the minimum you require, and then ifdef out the new code if the req is not met.

Unless we want to increase the minimum gnutls version required and cut off Centos/rhel 7? But IMHO it doesn't make sense yet, for an optional component.

On top of that, what's the reason for importing the whole gnutls headers wholesale into the source tree? Doesn't look right to me?

ssahani commented:
Can you come up with a fix

ssahani commented:
@keszybz  how to fix the meson?

bluca commented:
> Can you come up with a fix

You can add something like this in meson.build:

```
diff --git a/meson.build b/meson.build
index 73e47202bc..b07d01f875 100644
--- a/meson.build
+++ b/meson.build
@@ -1309,6 +1309,10 @@ else
 endif
 conf.set10('HAVE_DBUS', have)
 
+# NTS support in timesyncd requires new APIs compared to the minimum supported version
+have_gnutls_for_nts = (conf.get('HAVE_GNUTLS') == 1 and libgnutls.version().version_compare('>= 3.6.3'))
+conf.set10('GNUTLS_FOR_NTS', have_gnutls_for_nts)
+
 default_dnssec = get_option('default-dnssec')
 if skip_deps
         default_dnssec = 'no'
```

And then you can use \#if GNUTLS_FOR_NTS in the source files.

dnicolodi commented:
From some previous discussions I had the impression that systemd trying to consolidate on OpenSSL for cryptographic functionality these days.

bluca commented:
> From some previous discussions I had the impression that systemd trying to consolidate on OpenSSL for cryptographic functionality these days.

Any reference to that? I could only find https://github.com/systemd/systemd/issues/9531 which does not appear to have definitive conclusions one way or the other

dnicolodi commented:
> > From some previous discussions I had the impression that systemd trying to consolidate on OpenSSL for cryptographic functionality these days.
> 
> Any reference to that? I could only find #9531 which does not appear to have definitive conclusions one way or the other

I don't think there is something definitive written anywhere, but from time to time when size of dependencies comes up this is brought up as a possible thing to work on. I think work on this has been hold back by the licensing situation, but with OpenSL 2.0 on its way and with the fact that Debian now considers OpenSSL a system library, this should not be an issue anymore. @poettering may have a more official point of view

mbiebl commented:
> > From some previous discussions I had the impression that systemd trying to consolidate on OpenSSL for cryptographic functionality these days.
> 
> Any reference to that? I could only find #9531 which does not appear to have definitive conclusions one way or the other

I think the most recent discussion is at
https://github.com/systemd/systemd/pull/14743#issuecomment-582295118

I would like to add here, that at least for Debian the situation has changed in the mean time:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=924937#105
`consider OpenSSL, cups and libgcc as system libraries (re GPL-2) `

I very much appreciate, that @keszybz was very considerate in that regard. 
From Debian's POV there is nothing standing in the way anymore to go all OpenSSL if so desired.

bluca commented:
> > > From some previous discussions I had the impression that systemd trying to consolidate on OpenSSL for cryptographic functionality these days.
> > 
> > 
> > Any reference to that? I could only find #9531 which does not appear to have definitive conclusions one way or the other
> 
> I think the most recent discussion is at
> [#14743 (comment)](https://github.com/systemd/systemd/pull/14743#issuecomment-582295118)
> 
> I would like to add here, that at least for Debian the situation has changed in the mean time:
> 
> https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=924937#105
> `consider OpenSSL, cups and libgcc as system libraries (re GPL-2) `
> 
> I very much appreciate, that @keszybz was very considerate in that regard.
> From Debian's POV there is nothing standing in the way anymore to go all OpenSSL if so desired.

Nice! If we can switch entirely then we should do so, it sounds like a good change. Hopefully that linked PR can be updated and brought forward.

@ssahani would it be doable to switch this PR to openssl, if the core maintainers agree?

ssahani commented:
No you can are welcome to add ooenssl version.

ssahani commented:
@wingel  I tested with your server .can you confirm with the results.

wingel commented:
> @wingel I tested with your server .can you confirm with the results.

We don't log the NTS traffic and have a couple of hundred users of NTS so I can't really tell if you were successful or not.

Did you manage to get time from our servers?  If you used "nts.netnod.se" you are supposed to use "ntsts.sth.netnod.se" port 123 as NTPv4 Server.

ssahani commented:
> > @wingel I tested with your server .can you confirm with the results.
> 
> We don't log the NTS traffic and have a couple of hundred users of NTS so I can't really tell if you were successful or not.
> 
> Did you manage to get time from our servers? If you used "nts.netnod.se" you are supposed to use "ntsts.sth.netnod.se" port 123 as NTPv4 Server.

```

sudo killall -9 sudo SYSTEMD_LOG_LEVEL=debug ./systemd-timesyncd;sudo SYSTEMD_LOG_LEVEL=debug ./systemd-timesyncd
[1]    296490 killed     sudo killall -9 sudo SYSTEMD_LOG_LEVEL=debug ./systemd-timesyncd
SYSTEMD_LOG_LEVEL=debug: no process found
./systemd-timesyncd: no process found
Bus bus-api-timesync: changing state UNSET → OPENING
sd-bus: starting bus bus-api-timesync by connecting to /run/dbus/system_bus_socket...
Bus bus-api-timesync: changing state OPENING → AUTHENTICATING
Registering bus object implementation for path=/org/freedesktop/LogControl1 iface=org.freedesktop.LogControl1
Selected ntske server nts.netnod.se.
Added new ntske server nts.netnod.se.
systemd-timesyncd running as pid 296493
Resolving nts.netnod.se...
Bus bus-api-timesync: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RequestName cookie=2 reply_cookie=0 signature=su error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.1024 path=n/a interface=n/a member=n/a cookie=4294967295 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus bus-api-timesync: changing state HELLO → RUNNING
Got message type=signal sender=org.freedesktop.DBus.Local destination=n/a path=/org/freedesktop/DBus/Local interface=org.freedesktop.DBus.Local member=Connected cookie=4294967295 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.DBus destination=:1.1024 path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired cookie=4294967295 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.1024 path=n/a interface=n/a member=n/a cookie=4294967295 reply_cookie=2 signature=u error-name=n/a error-message=n/a
Requested service name already owned, failing connection.
Bus bus-api-timesync: changing state RUNNING → CLOSING
Bus bus-api-timesync: changing state CLOSING → CLOSED
Resolved ntske address 194.58.202.203:123 for nts.netnod.se.
Resolved ntske address [2a01:3f7:2:202::203]:123 for nts.netnod.se.
Selected address 194.58.202.203:123 of ntske server nts.netnod.se.
Connecting to ntske server 194.58.202.203:4460 (nts.netnod.se).
Sending request to to ntske server 194.58.202.203:4460 (nts.netnod.se).
Received '17258' bytes from NTSKE server
NTSKE: Received Next Protocol='0', AEAD Algorithm='15', Server='ntsts.sth.ntp.se', N Cookies='8' of size='100' each
Selected server ntsts.sth.ntp.se.
Resolving ntsts.sth.ntp.se...
Resolved address 194.58.202.202:123 for ntsts.sth.ntp.se.
Resolved address [2a01:3f7:2:202::202]:123 for ntsts.sth.ntp.se.
Selected address 194.58.202.202:123 of server ntsts.sth.ntp.se.
Connecting to time server 194.58.202.202:123 (ntsts.sth.ntp.se).
Sent NTS request to 194.58.202.202:123 (ntsts.sth.ntp.se).
NTS: Received responce message size '332' from server
NTS: Successfully verified UID received from server
NTS: Received AEEF message from server
NTS: Successfully decrypted AEEF message received from server
NTS: Extracted cookie of size '100' from AEEF message
NTP response:
  leap         : 0
  version      : 4
  mode         : 4
  stratum      : 2
  precision    : 0.000000 sec (-25)
  root distance: 0.000069 sec
  reference    : n/a
  origin       : 1607346721.119
  receive      : 1607346721.253
  transmit     : 1607346721.253
  dest         : 1607346721.288
  offset       : +0.049 sec
  delay        : +0.169 sec
  packet count : 1
  jitter       : 0.000
  poll interval: 64
  adjust (slew): +0.049 sec
  status       : 8193 sync
  time now     : 1607346721.288
  constant     : 2
  offset       : +0.049 sec
  freq offset  : -953982 (-14 ppm)
interval/delta/delay/jitter/drift 64s/+0.049s/0.169s/0.000s/-14ppm
Initial synchronization to time server 194.58.202.202:123 (ntsts.sth.ntp.se).

```

wingel commented:
> NTP response:
>   leap         : 0
>   version      : 4
>   mode         : 4
>   stratum      : 2
>   precision    : 0.000000 sec (-25)
>   root distance: 0.000069 sec
>   reference    : n/a
>   origin       : 1607346721.119
>   receive      : 1607346721.253
>   transmit     : 1607346721.253
>   dest         : 1607346721.288
>   offset       : +0.049 sec
>   delay        : +0.169 sec
>   packet count : 1
>   jitter       : 0.000
>   poll interval: 64
>   adjust (slew): +0.049 sec
>   status       : 8193 sync
>   time now     : 1607346721.288
>   constant     : 2
>   offset       : +0.049 sec
>   freq offset  : -953982 (-14 ppm)
> interval/delta/delay/jitter/drift 64s/+0.049s/0.169s/0.000s/-14ppm
> Initial synchronization to time server 194.58.202.202:123 (ntsts.sth.ntp.se).
> ```



> NTP response:
>   leap         : 0
>   version      : 4
>   mode         : 4
>   stratum      : 2
>   precision    : 0.000000 sec (-25)
>   root distance: 0.000069 sec
>   reference    : n/a
>   origin       : 1607346721.119
>   receive      : 1607346721.253
>   transmit     : 1607346721.253
>   dest         : 1607346721.288
>   offset       : +0.049 sec
>   delay        : +0.169 sec
>   packet count : 1
>   jitter       : 0.000
>   poll interval: 64
>   adjust (slew): +0.049 sec
>   status       : 8193 sync
>   time now     : 1607346721.288
>   constant     : 2
>   offset       : +0.049 sec
>   freq offset  : -953982 (-14 ppm)
> interval/delta/delay/jitter/drift 64s/+0.049s/0.169s/0.000s/-14ppm
> Initial synchronization to time server 194.58.202.202:123 (ntsts.sth.ntp.se).

Looks good to me.

bluca commented on src/timesync/meson.build:
```diff
@@ -6,6 +6,14 @@ systemd_timesyncd_sources = files('''
+        timesyncd-ntske-client.c
+        timesyncd-ntske-client.h
+        timesyncd-ntske-protocol.c
+        timesyncd-ntske-protocol.h
```
I would suggest to move all the new files in an optional meson include - that way you can remove all the \#ifdef from those sources, since there's really a lot of them, and use \#ifdef only in the common source files. Code will be much more readable and more robust that way.

Something like this should do the trick:

```
--- a/src/timesync/meson.build
+++ b/src/timesync/meson.build
@@ -13,6 +13,15 @@ systemd_timesyncd_sources = files('''
         timesyncd-server.h
 '''.split())
 
+if conf.get('GNUTLS_FOR_NTS') == 1
+        systemd_timesyncd_sources += files(['src/timesync/timesyncd-nts-client.c',
+                                            'src/timesync/timesyncd-nts-client.h',
+                                            'src/timesync/timesyncd-ntske-client.c',
+                                            'src/timesync/timesyncd-ntske-client.h',
+                                            'src/timesync/timesyncd-ntske-protocol.h',
+                                            'src/timesync/timesyncd-ntske-protocol.h'])
+endif
+
 timesyncd_gperf_c = custom_target(
         'timesyncd-gperf.c',
         input : 'timesyncd-gperf.gperf',
```

ssahani replied:
Yeah thanks for the review.

bluca commented on src/timesync/timesyncd-manager.c:
```diff
@@ -913,6 +1148,10 @@ void manager_flush_server_names(Manager  *m, ServerType t) {
                         server_name_free(m->fallback_servers);
+
+        if (t == SERVER_NTSKE)
+           while (m->ntske_servers)
```
nit: indentation is off

bluca commented on src/timesync/timesyncd-manager.c:
```diff
@@ -127,13 +131,33 @@ static int manager_send_request(Manager *m) {
+                        log_debug_errno(errno, "Sending NTS request to %s (%s) failed: %m", strna(pretty), m->current_server_name->string);
+                        return manager_connect(m);
+                }
+#endif
```
In this \#ifdef and more in the common files, shouldn't there be \#else branch to handle those cases? In other words, while the code now compiles without the required newer version of gnutls, what happens when running it?

ssahani commented:
Coming back to this . I did some work on openssl https://github.com/ssahani/systemd/pull/new/nts-timesync-openssl.  This fetches the cookies etc. But for the auth stuff I am not sure how to do this.  @poettering thoughts ?

poettering commented:
hmm, we typically prefer openssl these days over gnutls. And in light of this: https://lists.freedesktop.org/archives/systemd-devel/2020-December/045693.html — we should probably not do gnutls at all anymore, but go openssl only.

bluca commented:
Adding back the needs-rework label, due to the decision to go to openssl only

poettering commented:
I guess we can close this, and continue discussion in #17943
